### PR TITLE
Introduce *IfNecessary Load Actions

### LINKED ISSUE #144: Feature: Add new load paths to load only if data has not previously been loaded
When loading entities with any of the existing load actions, the loads will always occur and will always call the API. There may be times, such as when loading lookup tables, when the load should only occur if data has not been loaded previously. 

This proposal recommends introducing a new set of actions for loads, alongside the existing actions, that would only actually be dispatched if no data has been previously loaded for the specified entity.

A new set of actions is proposed to maintain backwards compatibility and function, method and constructor signature standard with existing actions. Modifying existing load actions would be possible by adding a new parameter to allow the developer to flag whether they want to load only if data has not been previously loaded. That would introduce a breaking signature change to all of the load actions and related facade methods. A new set of actions ensures the new functionality can be introduced without breaking any existing use cases. 

The proposed actions:

```
LoadIfMissing
LoadAllIfMissing
LoadManyIfMissing
```

Only **initiating** actions should be necessary, as these new actions would be filtered in the corresponding effects if the `loadedAt` timestamp for the given entity is not _nullish_, thus preventing the load from actually occurring. Further, these new actions should in turn dispatch their corresponding sibling actions, rather than calling the entity service directly. This will ensure that the existing load flows will be utilized if the `loadedAt` timestamps are indeed _nullish_.

### PR DESCRIPTION
**Pull request by jrista**
feat: add load if necessary action variants
    
     + Add LoadIfNecessary action
     + Add LoadAllIfNecessary action
     + Add LoadManyIfNecessary action
     + Add LoadPageIfNecessary action
     + Add LoadRangeIfNecessary action
     * Export new actions from public api
     + Add new operators for if necessary loads
     + Add new effects for if necessary loads
     * Include new effects in standard modules
     * Export new effects from public api
     + Add stateNameOfEntity utility function
     * Update builders to use new entityStateName function
    
Issue #144

-----

feat: add `defaultMaxAge` to entity operator
    
     * Introduce to `defaultMaxAge` to IEntityOptions
     + Add EntityAge with predefined common ages for use with `defaultMaxAge`
     * Update @Entity decorator to support passing name in only or options or both
     > Move key decorator tokens into entity-tokens.ts
     > Extract entity decorator into its own file entity-decorator.ts
     > Extract key decorator into its own file key-decorator.ts
     * Update index/public api to reference all exports from new locations
     * Added EntityAge enum to public api
     * Removed all imports from '../..'
    
Issues #144 #141

-----

feat: reorg action files and introduce *IfNecessary load actions

     * Refactor all actions into more files for easier management
       * Loads into load-*-actions.ts
       * CUURDS into create|update|upsert|replace|delete-actions.ts
       * DeleteByKeys into delete-by-key-actions.ts
       * Edits into edit-actions.ts
       * Selections into selection-actions.ts
       * Deselections into deselection-actions.ts
     + Add new *IfNecessary loader actions for each type of load
     + Add tsdoc to all load actions
     * Update library index/public api to reference all actions from new files
     * Update all otehr action references throughout library
     ^ Bump library version to 0.5.0-beta.3
    
Phase #1: This is an initial refactoring of action files for the if-necessary work
  
Issue #144

### CODE REVIEW

**Comment by jrista** [comment]
#### NOTE: DO NOT MERGE!! ####

This PR is just phase one of the work for this feature. Additional work is being done to integrate the new *IfNecessary actions into the effects. Do not merge until the second phase of the work has been committed.

**Comment by jrista** [comment]
It looks like our performance tests are failing. I'm a little concerned, as when I originally created those, the actual run times were well below (fractions of) the actual times I "expected". For example, the 50ms time for the 10k load was about 5x longer than the actual results I was getting at the time, which was ~10ms on average, and the 2000ms for the 1m load was about 2-3x longer. We are now topping those times consistently... I think I'd like to figure out why before we merge this. I would hate to hurt our performance numbers that much.

**Comment by jrista** [comment]
I'm starting to think the performance difference might in part be due to some changes in the jest libraries. I upgraded jest a minor version, and I even recall at the time that I had some issues after just with the normal run times of the full set of unit tests. 

When I run the performance tests within WebStorm, they all succeed consistently, however the run times for 10k and 100k are now about half the expected times, rather than ~1/5th the expected times, and the run time of the 1m test is close to the 2s expected time. 

It may be prudent to try and set up a lighter weight set of performance tests at some point...this might be testable just with node.js, as the reducer is just a function in the end.

**Comment by schuchard** [comment]
Seem's others are seeing issues at v24 and v25 https://github.com/facebook/jest/issues/7732

**Comment by jrista** [comment]
Interesting...I thought I'd only upgraded a minor version, but I think I upgraded jest previously as well, and that may have been one or even two major versions in that case. Well...its not a huge issue, although it may result in tests randomly failing. When I run locally in webstorm, I seem to be around half or so the specified thresholds for those performance tests. Not as good as they were originally, but better than I'm expecting.

File: projects/ngrx-auto-entity/src/index.ts [hunk 1]
```diff
@@ -30,109 +53,96 @@ export { fromEntityActions } from './lib/actions/action-operators';
...
+ * Entity Metadata Utilities
+ */
 export { makeEntity } from './lib/util/make-entity';
 export {
   nameOfEntity,
   pluralNameOfEntity,
   uriNameOfEntity,
+  stateNameOfEntity,
   entityComparer,
   entityTransforms
```
Comment (jrista) [inline projects/ngrx-auto-entity/src/index.ts:123]: Need to include `entityMaxAge`, which might better be renamed to `defaultMaxAge`

**Pull request merged**